File: 백준/Silver/15663. N과 M （9）/N과 M （9）.java
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;
import java.util.Arrays;

class Main {

    static StreamTokenizer sttk = new StreamTokenizer(new InputStreamReader(System.in));
    static StringBuilder sb = new StringBuilder();
    static int n;
    static int m;
    static int[] seq;
    static int[] ans;

    static int readInt() throws IOException {
        sttk.nextToken();
        return (int) sttk.nval;
    }

    public static void main(String[] args) throws IOException {
        n = readInt();
        m = readInt();
        seq = new int[n];
        ans = new int[m];

        for (int i = 0; i < n; i++) {
            seq[i] = readInt();
        }

        Arrays.sort(seq);

        travel(0, 0);

        System.out.println(sb);
    }

    static void travel(int idx, int visit) {
        if (idx == m) {
            for (int v : ans) {
                sb.append(v);
                sb.append(' ');
            }
            sb.append('\n');

            return;
        }

        int prev = -1;

        for (int i = 0; i < n; i++) {
            if (seq[i] == prev) continue;

            int mask = 1 << i;

            if ((visit & mask) == mask) continue;

            ans[idx] = seq[i];
            prev = seq[i];

            travel(idx + 1, visit | mask);
        }
    }
}

File: 백준/Silver/15663. N과 M （9）/N과 M （9）.kt
import java.io.StreamTokenizer

fun main() = with(StreamTokenizer(System.`in`.bufferedReader())) {
    val bw = System.out.bufferedWriter()
    val readInt = { nextToken(); nval.toInt() }

    val n = readInt()
    val m = readInt()

    val seq = IntArray(n) { readInt() }
    val stack = ArrayDeque<Int>(m)
    val cache = mutableSetOf<String>()

    seq.sort()

    fun travel(i: Int, visit: Int) {
        if (stack.size == m) {
            val s = stack.joinToString(" ")

            if (s !in cache) {
                cache += s
                bw.append("$s\n")
            }
            return
        }

        for (j in seq.indices) {
            val bit = 1 shl j

            if (visit and bit == 0) {
                stack.addLast(seq[j])
                travel(j, visit or bit)
                stack.removeLast()
            }
        }
    }

    for (i in seq.indices) {
        stack.addLast(seq[i])
        travel(i, 1 shl i)
        stack.removeLast()
    }

    bw.flush()
    bw.close()
}